### ISSUE TITLE
FreeBSD

### ISSUE THREAD

**Issue by xet7** [wekan/wekan#2662]
Moved to here from #1155 

From @dch 

As the above URL has disappeared, here's the original content -I'll see if I can get a more current version of wekan & FreeBSD to run based on it, & if so I'll see about making an official port and use a new issue for that.

# Install Wekan v0.32 on FreeBSD 11.0-RELEASE

## Prerequisites

Add user `wekan`

	# adduser

Install prerequisites packages

	# pkg install mongodb node4 npm2

Enable MongoDB

	# sysrc mongod_enable=YES
	# service mongod start

## Installation

As wekan

	$ cd
	$ fetch https://github.com/wekan/wekan/releases/download/v0.32/wekan-0.32.tar.gz
	$ tar -xzpf wekan-0.32.tar.gz

As root

	# chown -R wekan:wekan /home/wekan/bundle

As wekan

	$ cd ~/bundle/programs/server
	$ npm install
	$ npm install fibers

## Run

Considering that the default shell for a new user on FreeBSD is `/bin/sh`, the following ENV variables must be set according the following method before starting of Wekan. These must be adapted according the shell.

	$ MONGO_URL=mongodb://127.0.0.1:27017/wekan; export MONGO_URL
	$ ROOT_URL=https://example.com; export ROOT_URL
	$ MAIL_URL=smtp://user:pass@mailserver.example.com:25/; export MAIL_URL
	$ MAIL_FROM=wekan-admin@example.com; export MAIL_FROM
	$ PORT=8080; export PORT

Now it can be run

	$ cd ~/bundle
	$ node main.js

**Comment by xet7** [comment]
In general, you just download bundle as mentioned at:
https://github.com/wekan/wekan/wiki/Platforms

So there is wekan-3.xx.zip at https://releases.wekan.team

Required is:
- Node 8.16.0
- NPM
- MongoDB 3.2.x - 4.x

And start-wekan.sh is at https://github.com/wekan/wekan

I don't know how services etc work at FreeBSD.

**Comment by uropro** [comment]
# Install Wekan 3.45 on FreeBSD 12.0
The installation is possible but requires some small hacks: 
* The libraries 'phantomJS' is not available for FreeBSD 12.0, but as of now, it can be removed.
* The library 'bcrypt' causes troubles, but can be fixed.

**Install MongoDB 4.0:**

    # pkg install mongodb40 mongodb40-tools
    # sysrc mongod_enable=YES
    # service mongod start

**Install node.js 8:**

    # pkg install node8 npm-node8 bcrypt

**Make sure that python 2.7 will be found:**

    # cd /usr/local/bin
    # ln -s python2.7 python2

**Install wekan:**
Create an user 'wekan'

    # adduser

Fetch sources:

    # su wekan
    $ cd /home/wekan
    $ fetch https://releases.wekan.team/wekan-3.45.zip
    $ tar xzpf wekan-3.45.zip
    $ chown -R wekan:wekan /home/wekan/bundle

Remove phantomJS (so far wekan works well without phantomJS):

    $ cd ~/bundle/programs/server/npm/node_modules/meteor/lucasantoniassi_accounts-lockout/node_modules
    $ rm -rf phantomjs-prebuilt

Run npm first time (it will fail with bcrypt error):

    $ cd ~/bundle/programs/server
    $ npm install

Fix bcrypt problem:
```
npm install bcrypt
cd npm/node_modules
mv bcrypt ~/
cd ../..
npm install
```
now `npm install` should have completed successfully.
Clean up and install `fibers`:
```
mv ~/bcrypt npm/node_modules
npm install fibers
```
Start wekan, set environment (/bin/sh - style) first:
```
MONGO_URL=mongodb://127.0.0.1:27017/wekan; export MONGO_URL
ROOT_URL=http://YOURDOMAIN.com; export ROOT_URL
MAIL_URL=smtp://YOURSMTP.com:25/; export MAIL_URL
MAIL_FROM=wekan-admin@YOURDOMAIN.com; export MAIL_FROM
PORT=8080; export PORT

cd ~/bundle
node main.js
```

**Comment by xet7** [comment]
@uropro 

Would using `start-wekan.sh` also work? Because it has other settings too, like `WITH_API=true` that enabled API and exporting board.

**Comment by uropro** [comment]
@xet7 
yes, I think it could work. Setting WITH_API makes sense and other stuff in start-wekan.sh too.
Actually it would be even nicer to use the FreeBSD's rc.d mechanism and start wekan with something like `daemon -p ${pidfile} node main.js`.

**Comment by uropro** [comment]
Here is a rc.d wekan script to start it. 

1. Create a user 'wekan' without login, home directory /home/wekan for example. 
2. Copy all settings of the start-wekan.sh script found in the wekan repository into /home/wekan/wekan.conf.
3. Edit the /usr/local/etc/rc.d/wekan script:

```
#!/bin/sh
# PROVIDE: wekan
# REQUIRE: mongod nginx
# BEFORE:
# KEYWORD: shutdown

. /etc/rc.subr

name="wekan"
rcvar="wekan_enable"
pidfile="/var/run/${name}.pid"

. /home/wekan/wekan.conf
cd /home/wekan/bundle
command="/usr/sbin/daemon"
command_args="-P ${pidfile} -u wekan -r /usr/local/bin/node main.js"

load_rc_config $name
: ${wekan_enable:="NO"}

run_rc_command "$1"
```
4. enable it with `sysrc wekan_enable=yes`
5. start wekan: `service wekan start`

**Comment by jshumakerpruitt** [comment]
Hello all,

I tried downloading from https://releases.wekan.team/wekan-3.45.zip but got a 404.  Then, I downloaded 3.45 from the github release https://github.com/wekan/wekan/archive/v3.45.tar.gz, but the directory layout didn't match the layout in @uropro 's comment above.  

For example `find . -type d -name programs` and `find . -type d -name npm` don't return anything.  

```
$ls -1 wekan-3.45/
CHANGELOG.md
CONTRIBUTING.md
Dockerfile
LICENSE
README.md
SECURITY.md
Stackerfile.yml
app.env
app.json
client
config
docker-compose.yml
find.sh
fix-download-unicode
helm
i18n
meta
models
nginx
openapi
openshift
package-lock.json
package.json
packages
public
rebuild-wekan.bat
rebuild-wekan.sh
releases
sandstorm-pkgdef.capnp
sandstorm.js
scalingo.json
server
settings.json
snap
snap-src
snapcraft.yaml
stacksmith
start-wekan.bat
start-wekan.sh
status-wekan.sh
stop-wekan.sh
torodb-postgresql
```

There is no npm directory

**Comment by xet7** [comment]
When looking at https://releases.wekan.team it seems newest is wekan-3.56.zip

Only newest release is available, because old versions have bugs that are already fixed in newest version.

**Comment by xet7** [comment]
As mentioned [at Platforms page](https://github.com/wekan/wekan/wiki/Platforms#not-exposed-to-internet-bundle-for-raspi-3-arm64-windows-and-any-nodemongo-cpu-architectures-no-automatic-updates-no-sandboxing), there is also [bash install script and automatic upgrade script](https://github.com/wekan/wekan-bash-install-autoupgrade).

**Comment by wimvandijck** [comment]
Similar with 3.86. There is no main.js or bundle folder.
npm install finished ok. 

The install script is linux only. 
There's no real install documentation in the tarball/zipfile. 
Can anyone confirm this works on FreeBSD?

**Comment by xet7** [comment]
@wimvandijck 

As can be seen from [Dockerfile](https://github.com/wekan/wekan/blob/master/Dockerfile#L9), current used Node version is v12.16.1, latest npm, and some MongoDB v3.2.x - v4.x

Bundle usage is similar like at https://github.com/wekan/wekan/wiki/Raspberry-Pi

Newest bundle is at https://releases.wekan.team , currently wekan-3.86.zip .

That .zip file has bundle directory.

Settings are done as environment variables before running `node main.js` in bundle directory.
- At [start-wekan.sh](https://raw.githubusercontent.com/wekan/wekan/master/start-wekan.sh), for example `export ROOT_URL='http://localhost:2000'`

More about ROOT_URL at https://github.com/wekan/wekan/wiki/Settings

There should be some similar way to set environment variable on script or service on FreeBSD.

Usually MongoDB is installed to port 27017. So that should be [changed](https://github.com/wekan/wekan/blob/master/start-wekan.sh#L10), for example:
```
export MONGO_URL='mongodb://127.0.0.1:27017/wekan'
```
Also this is required, to enable API and export board:
```
export WITH_API='true'
```
Do you get any errors when trying to start Wekan?

**Comment by uropro** [comment]
wekan 3.86 works with FreeBSD 12.1 when installed as described above: PhantomJS and bcrypt fail to compile - delete PhantomJS folder. Move bcrypt folder away somewhere temporarily, then run `npm install` and finally restore bcrypt folder. Later  on you will see a message about performance impact due to the use of non-compiled bcrypt code. I didn't notice a problem with it.

@xet7 : Is phantomjs used anywhere in wekan?

**Comment by xet7** [comment]
@uropro 

I don't know, it seems from https://phantomjs.org/ that "PhantomJS is a headless web browser scriptable with JavaScript." If you don't notice any problems, then I think it's not needed.

Binary bcrypt was removed because of security issue, so Wekan uses Javascript-based bcrypt. I have not tried to add binary bcrypt again. I think there is no noticeable difference.

**Comment by xet7** [comment]
In general, current servers, browsers etc are so fast, that noticing any performance issues is only with very large boards and very big amount of users. Usually some bugfix can improve those a lot, because it could be just that too much data is loaded to webbrowser at once. Only visible data needs to be loaded.

**Comment by uropro** [comment]
After approx one year of usage, no user complained about missing PhantomJS, so I guess it is not needed.
Regarding bcrypt: I have the impression that `npm install` tries to compile it, which is not needed as @xet7 writes,  but I might be wrong. Anyway, the procedure of temporarily moving the bcrypt folder away allows to run `npm install` successfully (and removing PhantomJS).

Great software anyway, thanks a lot, @xet7 !

**Comment by xet7** [comment]
Thanks @uropro :)

**Comment by wimvandijck** [comment]
Hey guys,

thanks all for quick feedback. 
It seems the confusion was my mistake. I had first downloaded the source tarball v3.86.tar.gz :)

I got the zipfile from https://releases.wekan.team/wekan-3.86.zip and now it's all good ;-). Thanks!

**Comment by kapfenho** [comment]
Thanks @xet7 and @uropro (and others of course)!

I've installed Wekan v4.17.0 along your description and it runs nicely.

I've extended the `rc.d` file, below you find my version.

I was wondering why there is no FreeBSD port and if I should create one.
Do you know of any efforts/experiences/monumental fails/etc I should be aware of...?

```
#!/bin/sh

# PROVIDE: wekan
# REQUIRE: mongod
# BEFORE:
# KEYWORD: shutdown

# wekan_enable (bool):          Set it to YES to enable wekan
#                               Default: NO
# wekan_user (str):             Username of process owner
#                               Default: www
# wekan_group (str):            Groupname of process owner
#                               Default: www
# wekan_title (str):            Title of node process
#                               Default: wekan-node
# wekan_logfile (str):          Path of log file
#                               Default: /var/log/wekan/wekan.log
# wekan_config (str):           Path of config file
#                               Default: /usr/local/etc/wekan.conf

. /etc/rc.subr

name="wekan"
rcvar="wekan_enable"

load_rc_config $name

: ${wekan_enable:="NO"}
: ${wekan_user:="www"}
: ${wekan_group:="www"}
: ${wekan_title:="wekan-node"}
: ${wekan_logfile:="/var/log/wekan/wekan.log"}
: ${wekan_config:="/usr/local/etc/wekan.conf"}

. ${wekan_config}

appdir="/usr/local/www/wekan/bundle"
pidfile="/var/run/${name}/${name}.pid"
start_precmd="wekan_precmd"
procname="/usr/local/bin/node"
command="/usr/sbin/daemon"
command_args="-t ${wekan_title} -p ${pidfile} -o ${wekan_logfile} ${procname} main.js"
wekan_chdir="${appdir}"

wekan_precmd()
{
        if [ ! -d $(dirname ${pidfile}) ]; then
                mkdir -p $(dirname ${pidfile})
                chown ${wekan_user}:${wekan_group} $(dirname ${pidfile})
        fi
        if [ ! -d $(dirname ${wekan_logfile}) ]; then
                mkdir -p $(dirname ${wekan_logfile})
                chown ${wekan_user}:${wekan_group} $(dirname ${wekan_logfile})
        fi
}

run_rc_command "$1"
```

**Comment by xet7** [comment]
@kapfenho 

Yes you should create FreeBSD port.

I have not even figured out how to install FreeBSD or NetBSD. Previous install info for FreeBSD has been contributed by those who know how FreeBSD works. Ports, packaging etc is very welcome for all new platforms.

**Comment by hhartzer** [comment]
@kapfenho have you taken a stab at this by chance? I've never done any kind of Node port for FreeBSD.

**Comment by kapfenho** [comment]
I'm currently not using Wekan and I didn't create a port yet.

However, I can help to create a port for the app, but for submitting it we need a maintainer. Since I don't use the app I would soon become a bad/lazy maintainer.

To all: contact me if you run this app on BSD and you would like to maintain the port
